test/addons-napi/test_threadsafe_function/test_basic.c

```diff
@@ -0,0 +1,189 @@
+
+  if (value_to_collect != NULL) {
+    status = napi_call_function(env, collect_results, collect_results, 1,
+        &value_to_collect, NULL);
+    if (status != napi_ok) {
+      napi_throw_error(env, NULL, "Failed to call return value collector");
+    }
+  }
+}
+
+static napi_status marshal_data(napi_env env,
+                                void* data,
+                                napi_value* recv,
+                                size_t argc,
+                                napi_value* argv) {
+  int value;
+  TestAsyncData* async = (TestAsyncData*)data;
+  napi_status status = napi_create_object(env, recv);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  value = async->value;
+  uv_sem_post(&async->may_write);
+
+  status = napi_create_int32(env, value, argv);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  return napi_ok;
+}
+
+static void BasicTestThread(void* thread_data) {
+  napi_threadsafe_function async = (napi_threadsafe_function)thread_data;
+  void* data = NULL;
+  TestAsyncData* test_data = NULL;
+  int value;
+  napi_status status = napi_get_threadsafe_function_data(async, &data);
+  if (status != napi_ok) {
+    napi_fatal_error("BasicTestThread", NAPI_AUTO_LENGTH,
+        "Failed to retrieve async function data", NAPI_AUTO_LENGTH);
+  }
+  test_data = (TestAsyncData*)data;
+
+  for (value = 0; value < 5; value++) {
+    uv_sem_wait(&test_data->may_write);
+    test_data->value = value;
+    status = napi_call_threadsafe_function(async);
+    if (status != napi_ok) {
+      napi_fatal_error("BasicTestThread", NAPI_AUTO_LENGTH,
+          "Failed to call async function", NAPI_AUTO_LENGTH);
+    }
+  }
+}
+
+static napi_value StartThread(napi_env env, napi_callback_info info) {
+  const char* error;
+  size_t argc = 2;
+  napi_value argv[2];
+  napi_status status =
+      napi_get_cb_info(env, info, &argc, argv, NULL, NULL);
+  if (status != napi_ok) {
+    error = "Failed to retrieve JS callback";
+    goto throw;
+  }
+
+  if (async_func) {
+    error = "There is already an async function in place";
+    goto throw;
+  }
+
+  TestAsyncData* async_data = malloc(sizeof(TestAsyncData));
+  if (async_data == NULL) {
+    error = "Failed to allocate memory for test data";
+    goto throw;
+  }
+  async_data->value = -1;
+
+  status = napi_create_reference(env, argv[1], 1, &async_data->collect_results);
+  if (status != napi_ok) {
+    error = "Failed to create reference to function that collects the "
+        "JS callback return values";
+    goto free_data;
+  }
+
+  status = napi_create_threadsafe_function(env, argv[0], async_data, 1,
+      marshal_data, process_result, &async_func);
+  if (status != napi_ok) {
+    error = "Failed to create async function";
+    goto delete_reference;
+  }
+
+  if (uv_sem_init(&async_data->may_write, 1) != 0) {
+    error = "Failed to initialize write sem";
+    goto delete_async_function;
+  }
+
+  if (uv_thread_create((uv_thread_t*)async_data, BasicTestThread,
```

Comment:
`&async_data->thread` - I've pointed out similar issues several times now and I'll stop doing that from now.  Just stop using this dangerous idiom from now on.